Support a `-c` option in the `pnpm dlx` command

### pnpm version: 7.17.0

### Code to reproduce the issue:

```shell
pnpx --package lolcatjs --package cowsay -c \
  'echo "$npm_package_name@$npm_package_version" | cowsay | lolcatjs'
```

### Expected behavior:

Should install packages and call command.

### Actual behavior:

```shell
$ pnpx -p lolcatjs -p cowsay -c \
>            'echo "$npm_package_name@$npm_package_version" | cowsay | lolcatjs'
/home/.local/share/pnpm/store/v3/tmp/dlx-1174922:
 ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/-p: Not Found - 404

This error happened while installing a direct dependency of /home/.local/share/pnpm/store/v3/tmp/dlx-1174922

-p is not in the npm registry, or you have no permission to fetch it.

An authorization header was used: Bearer npm_[hidden]
```

### Additional information:

 - `node -v` prints: v16.18.1
 - Windows, macOS, or Linux?: Linux

--- a/.changeset/olive-readers-drop.md
+++ b/.changeset/olive-readers-drop.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/plugin-commands-script-runners": minor
+"pnpm": minor
+---
+
+The `pnpm dlx` command supports the `--shell-mode` (or `-c`) option. When used, the script is executed by a shell [#5679](https://github.com/pnpm/pnpm/issues/5679).

--- a/exec/plugin-commands-script-runners/src/dlx.ts
+++ b/exec/plugin-commands-script-runners/src/dlx.ts
@@ -16,11 +16,16 @@ import { makeEnv } from './makeEnv'
 
 export const commandNames = ['dlx']
 
+export const shorthands = {
+  c: '--shell-mode',
+}
+
 export function rcOptionsTypes () {
   return {
     ...pick([
       'use-node-version',
     ], types),
+    'shell-mode': Boolean,
   }
 }
 
@@ -35,11 +40,17 @@ export function help () {
     descriptionLists: [
       {
         title: 'Options',
-
         list: [
           {
             description: 'The package to install before running the command',
             name: '--package',
+          },
+          {
+            description: 'If exist, runs file inside of a shell. \
+  Uses /bin/sh on UNIX and \\cmd.exe on Windows. \
+  The shell should understand the -c switch on UNIX or /d /s /c on Windows.',
+            name: '--shell-mode',
+            shortAlias: '-c',
           },
         ],
       },
@@ -52,6 +63,7 @@ export function help () {
 
 export type DlxCommandOptions = {
   package?: string[]
+  shellMode?: boolean
 } & Pick<Config, 'reporter' | 'userAgent'> & add.AddCommandOptions
 
 export async function handler (
@@ -87,8 +99,10 @@ export async function handler (
     ? command
     : await getBinName(modulesDir, await getPkgName(prefix))
   await execa(binName, args, {
+    cwd: process.cwd(),
     env,
     stdio: 'inherit',
+    shell: opts.shellMode ?? false,
   })
 }
 

--- a/exec/plugin-commands-script-runners/src/dlx.ts
+++ b/exec/plugin-commands-script-runners/src/dlx.ts
@@ -46,9 +46,7 @@ export function help () {
             name: '--package',
           },
           {
-            description: 'If exist, runs file inside of a shell. \
-  Uses /bin/sh on UNIX and \\cmd.exe on Windows. \
-  The shell should understand the -c switch on UNIX or /d /s /c on Windows.',
+            description: 'Runs the script inside of a shell. Uses /bin/sh on UNIX and \\cmd.exe on Windows.',
             name: '--shell-mode',
             shortAlias: '-c',
           },

--- a/exec/plugin-commands-script-runners/test/dlx.e2e.ts
+++ b/exec/plugin-commands-script-runners/test/dlx.e2e.ts
@@ -90,3 +90,19 @@ test('dlx should fail when the package has no bins', async () => {
     }, ['is-positive'])
   ).rejects.toThrow(/No binaries found in is-positive/)
 })
+
+test('dlx should work in shell mode', async () => {
+  prepareEmpty()
+
+  await dlx.handler({
+    ...DEFAULT_OPTS,
+    dir: path.resolve('project'),
+    storeDir: path.resolve('store'),
+    package: [
+      'is-positive',
+    ],
+    shellMode: true,
+  }, ['echo "some text" > foo'])
+
+  expect(fs.existsSync('foo')).toBeTruthy()
+})

